[Flang][OpenMP][Runtime] Minor Flang runtime for OpenMP AMDGPU modifications

agozillon opened the pull request:
We have some modifications downstream to compile the flang runtime for amdgpu using clang OpenMP, some more hacky than others to workaround (hopefully temporary) compiler issues. The additions here are the non-hacky alterations.

Main changes:
	* Create freestanding versions of memcpy, strlen and memmove, and replace std:: references with these so that we can default to std:: when it's available, or our own Flang implementation when it's not. * Wrap more bits and pieces of the library in declare target wrappers (RT_* macros). * Fix some warnings that'll pose issues with werror on, in this case having the namespace infront of variables passed to templates.

Another minor issues that'll likely still pop up depending on the program you're linking with is that abort will be undefined, it is perhaps possible to solve it with a freestanding implementation as with memcpy etc. but we end up with multiple definitions in this case. An alternative is to create an empty extern "c" version (which can be empty or forwrd on to the builtin).

Co-author: Dan Palermo Dan.Palermo@amd.com

agozillon committed:
[Flang][OpenMP] Minor Flang runtime for OpenMP AMDGPU

We have some modifications downstream to compile the flang runtime for amdgpu using clang OpenMP, some more
hacky than others to workaround (hopefully temporary) compiler issues. The additions here are the non-hacky
alterations.

Main changes:
	* Create freestanding versions of memcpy, strlen and memmove, and replace std:: references with
	  these so that we can default to std:: when it's available, or our own Flang implementation when
          it's not.
        * Wrap more bits and pieces of the library in declare target wrappers (RT_* macros).
        * Fix some warnings that'll pose issues with werror on, in this case having the namespace infront
          of variables passed to templates.

Another minor issues that'll likely still pop up depending on the program you're linking with is that abort will be
undefined, it is perhaps possible to solve it with a freestanding implementation as with memcpy etc. but we end up
with multiple definitions in this case. An alternative is to create an empty extern "c" version (which can be empty
or forwrd on to the builtin).

Co-author: Dan Palermo Dan.Palermo@amd.com

agozillon reviewed (commented):
Added some replies inline, I'll update the PR once the discussions have been concluded!

klausler commented on flang-rt/include/flang-rt/runtime/work-queue.h:
```diff
@@ -127,8 +127,8 @@ class Elementwise {
   std::size_t elementAt_{0};
-  SubscriptValue subscripts_[common::maxRank];
-  SubscriptValue fromSubscripts_[common::maxRank];
+  SubscriptValue subscripts_[maxRank];
```
What is the warning from clang?

agozillon commented on flang/include/flang/Runtime/freestanding-tools.h:
```diff
@@ -23,6 +23,16 @@
 
+#if !defined(STD_MEMSET_UNSUPPORTED) && \
+    (defined(__CUDACC__) || defined(__CUDA__)) && defined(__CUDA_ARCH__)
+#define STD_MEMSET_UNSUPPORTED 1
```
Sounds good, will do so on the next pass!

klausler reviewed (commented):
Are there other APIs from `std::` in the runtime that should be replaced?

agozillon commented on flang-rt/lib/runtime/assign.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "flang-rt/runtime/tools.h"
 #include "flang-rt/runtime/type-info.h"
 #include "flang-rt/runtime/work-queue.h"
+#include "flang/Runtime/stop.h"
```
Will double check if the includes are necessary!

agozillon replied:
No requirement for this one anymore, thank you for pointing it out!

agozillon commented:
> Are there other APIs from `std::` in the runtime that should be replaced?

Not that we've found so far, but there will likely be others as the runtime gets built on! So we're happy to address them as they come up.

There is an issue with abort (C abort I think, rather than std::) currently, but the method we use to work around it downstream at the moment isn't particularly upstreamable, we create an local (to the .cpp source file with the issue) extern "C" function in-place of abort to forward to the fortran runtime's abort. Unfortunately, couldn't replicate similar functionality to the memset/memcpy in the PR as we end up with duplicate symbols.

agozillon commented on flang-rt/lib/runtime/copy.cpp:
```diff
@@ -10,6 +10,7 @@
 #include "stack.h"
 #include "flang-rt/runtime/descriptor.h"
 #include "flang-rt/runtime/terminator.h"
+#include "flang-rt/runtime/tools.h"
```
in this case it is unfortunately, due needing the memcpy from the fortran::runtime.

klausler replied:
But memcpy is defined in freestanding-tools.h, no?

agozillon commented:
Updated the PR stack, the main addition was new RT_OFFLOAD_API_GROUP wrappers in temporary-stack.cpp.

But otherwise I rebased and reformatted the PR to fix the code formatting!

But the PR should address all current comments and pass the CI now, please do feel free to add more comments however if you wish anything further changed! :-)

klausler commented on flang-rt/lib/runtime/extensions.cpp:
```diff
@@ -204,8 +204,8 @@ void FORTRAN_PROCEDURE_NAME(getarg)(
   if (length >= 1 && getlogin_r(arg, length) == 0) {
-    auto loginLen{std::strlen(arg)};
-    std::memset(
+    auto loginLen{Fortran::runtime::strlen(arg)};
```
Just `runtime::` is sufficient.

klausler commented on flang-rt/lib/runtime/temporary-stack.cpp:
```diff
@@ -97,7 +99,7 @@ void DescriptorStorage<COPY_VALUES>::resize(size_type newCapacity) {
   // behavior.
   if (data_ != nullptr) {
-    memcpy(newData, data_, capacity_ * sizeof(Descriptor *));
+    Fortran::runtime::memcpy(newData, data_, capacity_ * sizeof(Descriptor *));
```
Just `runtime::` suffices.

klausler commented on flang-rt/lib/runtime/time-intrinsic.cpp:
```diff
@@ -276,13 +276,13 @@ static void DateAndTimeUnavailable(Fortran::runtime::Terminator &terminator,
     const Fortran::runtime::Descriptor *values) {
   if (date) {
-    std::memset(date, static_cast<int>(' '), dateChars);
+    Fortran::runtime::memset(date, static_cast<int>(' '), dateChars);
```
Just `runtime::` is needed here and in the following lines.

klausler commented on flang/include/flang/Runtime/freestanding-tools.h:
```diff
@@ -79,7 +98,52 @@ fill_n(A *start, std::size_t count, const B &value) {
+  while (count--) {
+    *to++ = *from++;
+  }
+  return;
```
You don't need a `return;` statement before the closing brace of a function.

agozillon commented:
Last PR should address all the comments currently. I'll git squash and clang-format to make the CI happy once we're happy with the PR (so as to make reviewing any changes easier for the time being).

vzakhari commented on flang-rt/lib/runtime/time-intrinsic.cpp:
```diff
@@ -420,7 +423,7 @@ static void GetDateAndTime(Fortran::runtime::Terminator &terminator, char *date,
       [&](char *dest, std::size_t destChars, std::size_t len) {
         auto copyLen{std::min(len, destChars)};
-        std::memcpy(dest, buffer, copyLen);
+        Fortran::runtime::memcpy(dest, buffer, copyLen);
```
```suggestion
        runtime::memcpy(dest, buffer, copyLen);
```

vzakhari reviewed (approved):
LGTM, with one minor comment.  Thank you!
